### PR TITLE
Deprecate `MemoryPool`

### PR DESCRIPTION
**Pull request by dalg24**
Following up on #7393

### CODE REVIEW

**Review by masterleinad — CHANGES_REQUESTED** [review]
```
make[2]: *** No rule to make target `Kokkos_MemoryPool.o', needed by `libkokkos.a'.  Stop.
```
You have to remove `Kokkos_MemoryPool.cpp` from `KOKKOS_SRC`.

**Comment by dalg24** [comment]
I need to do something like https://github.com/kokkos/kokkos/pull/7432/commits/7e0fc5e296b778e06d49ea40810aaa2f6b108359#diff-8b14e0420249c1698d04b2eb3998f490bdd7ec0fa2c24eb02572e71405997402R1627-R1631 ?

**Comment by masterleinad** [comment]
> I need to do something like [7e0fc5e#diff-8b14e0420249c1698d04b2eb3998f490bdd7ec0fa2c24eb02572e71405997402R1627-R1631](https://github.com/kokkos/kokkos/commit/7e0fc5e296b778e06d49ea40810aaa2f6b108359#diff-8b14e0420249c1698d04b2eb3998f490bdd7ec0fa2c24eb02572e71405997402R1627-R1631) ?

Yes, `filter-out` didn't work for me (or I couldn't get the syntax right).

File: core/unit_test/TestOther.hpp [hunk 1]
```diff
@@ -16,7 +16,10 @@
 
 #ifndef KOKKOS_TEST_OTHER_HPP
 #define KOKKOS_TEST_OTHER_HPP
+
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
 #include <TestMemoryPool.hpp>
+#endif
```
Comment (masterleinad) [inline core/unit_test/TestOther.hpp:22]: You don't have to disable deprecation warnings in that file?

File: core/perf_test/CMakeLists.txt [hunk 2]
```diff
@@ -128,7 +128,7 @@ kokkos_add_benchmark(PerformanceTest_Benchmark SOURCES ${BENCHMARK_SOURCES})
 kokkos_add_benchmark(Benchmark_Atomic_MinMax SOURCES test_atomic_minmax_simple.cpp)
 
 # FIXME_NVHPC
-if(NOT KOKKOS_CXX_COMPILER_ID STREQUAL NVHPC)
+if(NOT KOKKOS_CXX_COMPILER_ID STREQUAL NVHPC AND Kokkos_ENABLE_DEPRECATED_CODE_4)
   kokkos_add_benchmark(PerformanceTest_Mempool SOURCES test_mempool.cpp)
 endif()
```
Comment (masterleinad) [inline core/perf_test/CMakeLists.txt:133]: You don't have to disable deprecation warnings in `test_mempool.cpp`?

Comment (dalg24) [reply core/perf_test/CMakeLists.txt:133]: Yes I do 
```
/var/jenkins/workspace/Kokkos_PR-7431/core/perf_test/test_mempool.cpp:33:28: error: 'template<class Device> class Kokkos::MemoryPool' is deprecated [-Werror=deprecated-declarations]
 using MemoryPool = Kokkos::MemoryPool<ExecSpace>;
                            ^~~~~~~~~~
```
I find it a bit scary that only the GCC 8.4 build on Jenkins failed

**Review by masterleinad — APPROVED** [review]

**Comment by dalg24** [comment]
I looks like it is used in Sacado...
```
packages/sacado/src/Sacado_DynamicArrayTraits.cpp:15:    const Kokkos::MemoryPool<Kokkos::OpenMP>* global_sacado_openmp_memory_pool = 0;
packages/sacado/src/Sacado_DynamicArrayTraits.cpp:23:    const Kokkos::MemoryPool<Kokkos::Cuda>* global_sacado_cuda_memory_pool_host = 0;
packages/sacado/src/Sacado_DynamicArrayTraits.cpp:24:    const Kokkos::MemoryPool<Kokkos::Cuda>* global_sacado_cuda_memory_pool_device = 0;
packages/sacado/src/Sacado_DynamicArrayTraits.cpp:26:    __device__ const Kokkos::MemoryPool<Kokkos::Cuda>* global_sacado_cuda_memory_pool_on_device = 0;

**Comment by crtrott** [comment]
Ok there is usage out there. And it doesn't look like it has much of a crazy interface I guess. So maybe just keep it around, make sure our testing is good enough and document it properly ...

**Pull request closed without merging**